### DISCUSSION TITLE
ZITADEL as LDAP Server

### DISCUSSION THREAD

**Discussion by fforootd** [zitadel/zitadel#1929] [Ideas]
Sometime it would be interesting to emulate a LDAP Interface for external clients so that they can interface with ZITADEL over classic LDAP. 

For example atlassian tools rely on LDAP.

However this is not yet scheduled.

**Comment by gc-ss** [comment]
Requests were made at https://gitter.im/hashicorp-nomad/Lobby?at=610049306ccf813819decabc

```
Florian Apolloner @apollo13
mhm the docs don't seem to mention ldap, any chance of using that as user source?
```

**Comment by sueess** [comment]
Is this still a topic? Some of our client of our SaaS solution have their own user directory which uses ldap. What are the possibilities to connect an organization with ldap. The goal would be that we do not need an additional IAM implementation. Apart from LDAP, what are the general possibilities to integrate with third party IAM solutions?

**Reply by fforootd** [reply]
Hi @sueess 

If I understand you correctly you want to connect ZITADEL to an existing LDAP/AD, correct? 
In that case the discussion https://github.com/zitadel/zitadel/discussions/1938 is for you.

Here it is more worth discussing if ZITADEL should act as an LDAP server itself.

**Comment by Fudster** [comment]
I would love to see this, A lot of the software I use has LDAP as external auth

**Comment by jory3** [comment]
That would allow us to switch to Zitadel: We have many linux hosts on which we'd like to do user auth using Zitadel.

**Comment by nwmcsween** [comment]
For others looking casdoor seems to offer this functionality.

**Reply by catholic-indulgence-vaper** [reply]
Only auth with login/pass - no groups, no email fields, etc.

**Comment by brandan-schmitz** [comment]
I would really like to see this feature added as well. A specific use case for example is being able to login to a Nexus OSS registry instance. Their free tier only has support for LDAP as an external authentication method, and I would love to be able to use my zitadel instance to authenticate users to that. It would need the ability to pass user authorizations so that on the nexus side you can configure user and group mappings based off the roles.

There are many many other pieces of software where this would be useful as well, this is just an example one that could be easily tested against given that Nexus has a free version.

**Comment by julien-fremont** [comment]
I work in a SaaS company where we sell environments based on our partners' software (mainly HCL and IBM). We want to provide our customers with a unified user management interface as well as capabilities to integrate with their SSO providers if requested.

Zitadel seems well suited for this use case, but it cannot act as an LDAP server as this discussion suggests. Some of the tools we offer require an LDAP interface to map users and groups to tool roles, although they can delegate the login process via SAML or OAuth. As you can imagine, I can't just add alternative user and group sources to these software, so that the LDAP bridge is mandatory for our application.

Is LDAP server emulation planned at some point? I'm not very familiar with LDAP from a directory perspective, nor am I a developer. So I don't have a clear idea of how much work is needed to add this feature. But I might be able to convince my manager to appoint someone on the subject and do a PR if this work is not a 3 year project.

**Comment by helmut72** [comment]
Develope an own LDAP server is hard, but maybe an existing LDAP server can be used, if there is a backend module for Zitadels SQL (or API)?

https://glauth.github.io/docs/backends.html

**Reply by mf-in-mun** [reply]
I suppose that's a really good idea.

Have you already made post on their Discord?

**Reply by fforootd** [reply]
Yeah something like this could work if somebody invests time 😁 

At least on our end time is a constraint right now for this topic.

**Reply by mf-in-mun** [reply]
Totally understand.

I'm not a go guy. Sorry.

But hopefully time comes one day.

**Reply by helmut72** [reply]
Is it possible to access Zitadels database (Postgres) like this?

https://glauth.github.io/docs/databases.html
https://glauth.github.io/docs/databases.html#database-schema

**Comment by tbshfr** [comment]
Is this a feature that will be implemented in the future? I need it for SSSD, but I can wait. If it will not be implemented, I will use an LDAP server as an identity provider. However, if it will be implemented, it would not be necessary, and Zitadel would become my source of truth. I think this is the only feature that Zitadel currently lacks

**Reply by fforootd** [reply]
We currently lack the capacity to implement the LDAP server 😓  but we would certainly encourage the community to add LDAP 😁

**Reply by tbshfr** [reply]
Thank you for your quick response! Currently this is above my skill level, but maybe in the future... 😁

**Reply by DrBlury** [reply]
I would like to tackle this. I did not yet contribute to Zitadel but would love to get involved. But as others have mentioned, implementing an LDAP Server from scratch won't be an easy task. So it will take some time.

**Reply by fforootd** [reply]
I think it should be feasible to add an existing LDAP implemetation on top of our resources but starting can be challeging

**Comment by nwmcsween** [comment]
Look at casdoor, it's similar to zitadel and has ldap functionality

**Reply by fforootd** [reply]
I do not want sound "bad" here, but what always puts me of with casdoor is that its not really clear "who" is behind it.

And yes I am fully biased so take this with a LOT of salt 😁

**Reply by Codelica** [reply]
Authentik has [outposts](https://docs.goauthentik.io/docs/add-secure-apps/outposts/) to handle services like LDAP separately, which might be interesting.

**Reply by fforootd** [reply]
Yeah that is an interesting concept.

Since we already can consume ldap, which to me has more network related problems, I would say integrating it into our api code should not be "that" problematic.

Main thing here is that we really lack the capacity 😁

**Comment by ekozan** [comment]
https://github.com/alan-turing-institute/apricot  can be an solution

**Comment by Kidswiss** [comment]
I cobbled together a plugin for [Glauth](https://github.com/glauth/glauth): https://github.com/Kidswiss/zitadel-ldap

It's been running in my homelab for a few days so far. But it should be treated as a POC at best.
